desc/compute/_stability.py

```diff
@@ -231,3 +231,513 @@ def _magnetic_well(params, transforms, profiles, data, **kwargs):
+    )
+
+    B = B.at[i, l, j, k].set(jnp.sqrt(f[i, l, j + 1]) * (j - k == 0))
+    B_inv = B_inv.at[i, l, j, k].set(1 / jnp.sqrt(f[i, l, j + 1]) * (j - k == 0))
+
+    A_redo = B_inv @ A @ jnp.transpose(B_inv, axes=(0, 1, 3, 2))
+
+    w, _ = jnp.linalg.eigh(A_redo)
+
+    lam = jnp.real(jnp.max(w, axis=(2,)))
+
+    lam = jnp.max(lam)
+
+    data["ideal_ball_gamma2"] = lam
+
+    return data
+
+
+@register_compute_fun(
+    name="Newcomb_metric",
+    label="\\mathrm{Nwecomb-metric}",
+    units=" ",
+    units_long=" ",
+    description="A measure of Newcomb's distance from marginality",
+    dim=1,
+    params=["Psi"],
+    transforms={"grid": []},
+    profiles=[],
+    coordinates="rtz",
+    data=[
+        "a",
+        "g^aa",
+        "g^ra",
+        "g^rr",
+        "cvdrift",
+        "cvdrift0",
+        "|B|",
+        "B^zeta",
+        "p_r",
+        "phi",
+        "iota",
+        "iota_r",
+        "psi",
+        "psi_r",
+        "rho",
+    ],
+)
+def _Newcomb_metric(params, transforms, profiles, data, *kwargs):
+    """
+    Ideal-ballooning growth rate proxy.
+
+    This function uses a finite-difference method to integrate the
+    marginal stability ideal-ballooning equation
+
+    d/dζ(g dX/dζ) + c X = 0, g > 0
+
+    using the Newcomb's stability criterion
+
+    where
+
+    kappa = b dot grad b
+    g = a_N^3 * B_N * (b dot grad zeta) * |grad alpha|^2 / B,
+    c = a_N/B_N * (1/ b dot grad zeta) * dpsi/drho * dp/dpsi
+        * (b cross kappa) dot grad alpha/ B**2,
+    f = a_N * B_N^3 *|grad alpha|^2 / bmag^3 * 1/(b dot grad zeta),
+    are needed along a field line to solve the ballooning equation once.
+
+    To obtain the parameters g, c, and f, we need a set of parameters
+    provided in the list ``data`` above. Here's a description of
+    these parameters:
+
+    - a: minor radius of the device
+    - g^aa: |grad alpha|^2, field line bending term
+    - g^ra: (grad alpha dot grad rho) integrated local shear
+    - g^rr: |grad rho|^2 flux expansion term
+    - cvdrift: geometric factor of the curvature drift
+    - cvdrift0: geoetric factor of curvature drift 2
+    - |B|: magnitude of the magnetic field
+    - B^zeta: inverse of the jacobian
+    - p_r: dp/drho, pressure gradient
+    - psi_r: radial gradient of the toroidal flux
+    - phi: coordinate describing the position in the toroidal angle
+    along a field line
+
+    Here's how we define the Newcomb metric:
+    If zero crossing is at -inf (root finder failed), use the Y coordinate
+    as a metric of stability else use the zero-crossing point on the X-axis
+    as the metric
+    """
+    rho = data["rho"]
+
+    psi_b = params["Psi"] / (2 * jnp.pi)
+    a_N = data["a"]
+    # a_N is 0.4
+    B_N = 2 * psi_b / a_N**2
+
+    N_zeta0 = int(11)
+    # up-down symmetric equilibria only
+    zeta0 = jnp.linspace(-jnp.pi / 2, jnp.pi / 2, N_zeta0)
+
+    iota = data["iota"]
+    shat = -rho / iota * data["iota_r"]
```

Comment:
prefer if used existing registered compute quantity `shear` instead of defining the intermediate `shat`

Reply:
Done.